Allow override inputs to export recipe

guangy10 opened the pull request:
# What does this PR do?

Enable dynamism at seq_len dim in order to utilize parallel `prefill` in the executorch runtime. In this PR,
 - allow caller side to override example inputs, dynamic shapes and strict flag, but keep the default unchanged for BC
 - add unit test to cover export with dynamic shapes, and strict False as it's the mainstream in latest version of `torch.export`
 - make the unit test non-slow, to avoid being skipped on PRs and causing regressions
 - add test for `HybirdCache`

Tests
`pytest tests/utils/test_cache_utils.py -vv -s -k cache_exportability`
```
collected 23 items / 20 deselected / 3 selected

tests/utils/test_cache_utils.py::CacheExportIntegrationTest::test_dynamic_cache_exportability PASSED                                                             [ 33%]
tests/utils/test_cache_utils.py::CacheExportIntegrationTest::test_hybrid_cache_exportability PASSED                                                              [ 66%]
tests/utils/test_cache_utils.py::CacheExportIntegrationTest::test_static_cache_exportability PASSED                                                              [100%]
```


## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.  Needed by the downstream in [optimum-executorch](https://github.com/huggingface/optimum-executorch). https://github.com/huggingface/optimum-executorch/issues/53
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you add new tests? Yes


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @ydshieh

CC: @tugsbayasgalan

Commit:
Add option to specify dynamic shapes during export

ydshieh commented on src/transformers/integrations/executorch.py:
```diff
@@ -208,12 +210,19 @@ def convert_and_export_with_cache(
+        dynamic_shapes = (
+            dynamic_shapes
+            if dynamic_shapes is not None
+            else {"input_ids": {1: torch.export.Dim.AUTO}, "cache_position": None}
```
Could you explain a bit what is the value like `torch.export.Dim.AUTO` means, and the one in [this guide](https://pytorch.org/docs/stable/export.html) 

```
# Create a dynamic batch size
batch = Dim("batch")
# Specify that the first dimension of each input is that batch size
dynamic_shapes = {"x1": {0: batch}, "x2": {0: batch}}
```

IIRC, the key `0` or `1` means which dimension, but don't know `torch.export.Dim.AUTO` or `Dim("batch")`.

guangy10 replied:
>  key 0 or 1 means which dimension

Correct.

> torch.export.Dim.AUTO or Dim("batch")

It's new feature introduced since 2.6.0, smarter than the traditional way of specifying the dynamic range, e.g Dim("seq_len", min, max). @tugsbayasgalan can explain more. @ydshieh if you think `Dim("seq_len", min, max)` is easier to understand, we can switch to explicit dynamism.

ydshieh replied:
If you can find a doc that explains what `torch.export.Dim.AUTO` does and why it's smarter, adding the link along with a a short comment in the source code here, that would be nice. 

When I look [doc](https://pytorch.org/docs/stable/export.html#torch.export.export) [source](https://github.com/pytorch/pytorch/blob/v2.6.0/torch/export/__init__.py), I can't find any info.

But if you can't find neither, it's fine. In this case, maybe we can reflect this situation to pytorch team.

tugsbayasgalan replied:
Dim.AUTO works by refining the range as you encounter shape related constraints (possibly specializing to static integer). So in that sense, Dim.AUTO is less likely to run into shape related errors and doesn't require user code change as it will just respect user code intention. If you really want to keep seq_len to be dynamic, you should use Dim.DYNAMIC which will error when we specialize. For official doc, cc @pianpwk

ydshieh commented on src/transformers/integrations/executorch.py:
```diff
@@ -208,12 +210,19 @@ def convert_and_export_with_cache(
+        dynamic_shapes = (
+            dynamic_shapes
+            if dynamic_shapes is not None
+            else {"input_ids": {1: torch.export.Dim.AUTO}, "cache_position": None}
```
also, doing this way, it seems we remove the possibility of using static shape? Is this intended?

guangy10 replied:
True. We may always want the seq_len dim to be dynamic if we want to utilize prefill to speed up prompt token processing. If you are concerning about BC, I think I can remove this default and only set dynamism if it's passed via  arg, and add a test for it. wdyt?

ydshieh replied:
it's ok, we can go this way. If there are BC concerns reported, we can revise.

guangy10 replied:
Actually I think it's better not force everything to set dynamic shapes especially for older versions of torch. I will just move this out to tests

SunMarc commented on src/transformers/integrations/executorch.py:
```diff
@@ -208,12 +210,19 @@ def convert_and_export_with_cache(
+        dynamic_shapes = (
+            dynamic_shapes
+            if dynamic_shapes is not None
+            else {"input_ids": {1: torch.export.Dim.AUTO}, "cache_position": None}
```
as @ydshieh said, it would be nice to understand why you choose this default ? Also since you set `cache_position`  to None, we don't really need to specify it no ?

guangy10 replied:
Yeah, we want to set the seq_len dim to be dynamic in order to prefill prompt tokens. `cache_position` will be exported statically as it's None (same as not specified).

guangy10 replied:
> in this case, let's just remove cache_position maybe

Either the entire dynamic_shapes are None or you will need to be explicit about each input arg, otherwise you may get this:
```
E               torch._dynamo.exc.UserError: When `dynamic_shapes` is specified as a dict, its top-level keys must be the arg names ['input_ids', 'cache_position'] of `inputs`, but here they are ['input_ids']. Alternatively, you could also ignore arg names entirely and specify `dynamic_shapes` as a list/tuple matching `inputs`. For more information about this error, see: https://pytorch.org/docs/main/generated/exportdb/index.html#dynamic-shapes-validation
```

SunMarc commented on src/transformers/integrations/executorch.py:
```diff
@@ -182,15 +182,17 @@ def convert_and_export_with_cache(
     model: PreTrainedModel,
     example_input_ids: Optional[torch.Tensor] = None,
     example_cache_position: Optional[torch.Tensor] = None,
+    dynamic_shapes: Optional[dict] = None,
 ):
     """
     Convert a `PreTrainedModel` into an exportable module and export it using `torch.export`,
     ensuring the exported model is compatible with `ExecuTorch`.
 
     Args:
         model (`PreTrainedModel`): The pretrained model to be exported.
-        example_input_ids (`torch.Tensor`): Example input token id used by `torch.export`.
-        example_cache_position (`torch.Tensor`): Example current cache position used by `torch.export`.
+        example_input_ids (`Optional[torch.Tensor]`): Example input token id used by `torch.export`.
+        example_cache_position (`Optional[torch.Tensor]`): Example current cache position used by `torch.export`.
+        dynamic_shapes(`Optional[dict]`): Dynamic shapes used by `torch.export`.
 
     Returns:
         Exported program (`torch.export.ExportedProgram`): The exported program generated via `torch.export`.
```
can we add the link to torch.export docs somewhere in this docstring ? This will help understanding how to use dynamic_shapes for example. `https://pytorch.org/docs/stable/export.html`

guangy10 replied:
Done

SunMarc reviewed (commented):
Thanks ! left a few comments

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_37508). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

ydshieh commented:
Other than the nit comments I left, just a question: so this dynamism is necessary if the exported model want to be used to call prompts (i.e. prefill stage) as well as to use the generation steps (here seq=1)?

guangy10 commented on tests/utils/test_cache_utils.py:
```diff
@@ -283,6 +281,14 @@ def test_static_cache_exportability(self):
 
+        # Export with dynamic shapes using Dim.AUTO
+        tokenizer = AutoTokenizer.from_pretrained(model_id)
+        input_ids = tokenizer("Here's everything I know", return_tensors="pt").input_ids
+        dynamic_shapes = {"input_ids": {1: torch.export.Dim.AUTO}, "cache_position": None}
+        exported_program = convert_and_export_with_cache(
+            model, example_input_ids=input_ids, dynamic_shapes=dynamic_shapes
+        )
```
I'm export this model with a non-specialized `input_ids` and the dynamic shape on "seq_len" dim. @tugsbayasgalan Is there a way I can assert the exported program do have the dim set to be dynamic, and inspect what the dynamic range is?

ydshieh commented on src/transformers/integrations/executorch.py:
```diff
@@ -209,14 +214,21 @@ def convert_and_export_with_cache(
+                strict=strict if strict is not None else True,
             )
         else:
+            if dynamic_shapes is not None:
+                logging.warning("Dynamic shapes spec will be ingored for < 2.6.0.")
+            if strict is not None:
+                logging.warning("strict flag spec will be ingored for < 2.6.0.")
```
Hmm, first, I think we should have

if is_torch_greater_or_equal("2.6.0"):
   ...
elif is_torch_greater_or_equal("2.5.0"):
   ...
else:
  ...
right ...?


If so, the 2 new messages have to be in both `elif` and `else`.

Also, let's rephrase it as 

> Dynamic shapes spec will be ignored by `convert_and_export_with_cache` for < 2.6.0.

guangy10 replied:
That could be one option. Alternatively, we can simplify it by just not use dynamic shapes for torch<2.6 including 2.5, which is BC.

SunMarc commented on src/transformers/integrations/executorch.py:
```diff
@@ -209,14 +214,21 @@ def convert_and_export_with_cache(
         else:
+            if dynamic_shapes is not None:
+                logging.warning("Dynamic shapes spec will be ingored for < 2.6.0.")
+            if strict is not None:
+                logging.warning("strict flag spec will be ingored for < 2.6.0.")
```
let's default `strict` to `True`. Does it require 2.6 to set strict to False ?

guangy10 replied:
I don't have strong option to default True or False, but I think the export team wants to promote default to `False` in newer version of torch. cc: @tugsbayasgalan

guangy10 commented on src/transformers/integrations/executorch.py:
```diff
@@ -209,14 +214,21 @@ def convert_and_export_with_cache(
 
-        if is_torch_greater_or_equal("2.5.0"):
+        if is_torch_greater_or_equal("2.6.0"):
+            # `Dim.AUTO` is supported in torch >= 2.6.0.
```
Good catch

guangy10 commented:
> just a if/else v.s. if/elif/else question and 2 nits, but overall LGTM.

Shared my thoughts here: https://github.com/huggingface/transformers/pull/37508#discussion_r2047455608. Just in case you missed it. Let me know you thoughts

guangy10 commented:
More comments on this PR? Can we merge it if no?

guangy10 commented:
I will rebase this PR since https://github.com/huggingface/transformers/pull/37728 has been merged

ydshieh reviewed (approved):
Thank you for the iterations. LGTM, torched test is passing and fast to run 👍 

I will wait until the end of the day before merge to see if @SunMarc has any comment.

SunMarc reviewed (approved):
LGTM ! Please fix the conflits and we are good to merge

guangy10 commented:
> LGTM ! Please fix the conflits and we are good to merge

Couple updates:
  - Added a non-slow test to cover export for hybird cache
  - Rebased to latest trunk
  
@SunMarc @ydshieh Let me know if there are new comments

ydshieh commented:
Thanks
